staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -27,12 +27,66 @@ import (
 
 // DelayingInterface is an Interface that can Add an item at a later time. This makes it easier to
 // requeue items after failures without ending up in a hot-loop.
+// A Delaying queue can be thought of two separate queues that work together - an 'active' queue of
+// items which we want to process ASAP and a 'waiting' queue of items which must wait for a period of time
+// to elapse before they are popped from the 'waiting' queue and added to the 'active' queue.
 type DelayingInterface interface {
 	Interface
 	// AddAfter adds an item to the workqueue after the indicated duration has passed
 	AddAfter(item interface{}, duration time.Duration)
 }
 
+// EnhancedDelayingInterface adds methods to a DelayingInterface which allows clients finer control
+// over the delaying queue behaviours and the ability to inspect the state of the 'waiting' queue.
+type EnhancedDelayingInterface interface {
+	DelayingInterface
+	// AddWithOptions allows full control over the delaying queue features.
+	AddWithOptions(item interface{}, opts DelayingOptions)
+	// DoneWaiting will remove an item from the 'waiting' queue.
+	// If you want to cancel 'active' queued items you need to call Done().
+	DoneWaiting(item interface{})
+	// IsWaiting waits for fill and returns true if the item is found queued
+	// in the 'waiting' queue and how long it has to wait before it is ready.
+	IsWaiting(item interface{}) (bool, time.Duration)
+	// LenWaiting waits for fill and returns the length of the 'waiting' queue.
+	LenWaiting() int
+	// NextReady waits for fill and returns the item that is at the head of the 'waiting'
+	// queue and how long it has to wait before it is ready.
+	NextReady() (interface{}, time.Duration)
+	// WaitFill blocks until the delaying queue has drained all items from its feeder channel, processed
+	// them into their appropriate queues and updated the nextReadyAt time as necessary.
+	// If you are testing with a fake clock it is a good idea to call WaitFill BEFORE stepping the clock.
+	WaitFill()
```

Comment:
My expectation is that anyone wanting to write tests which involves a Delaying queue will need to do something like WaitFill before they can make any assumptions - or in the case I stress in the comments, adjust a fake clock used by the test and the queue (otherwise their tests will exhibit the same raciness issues that ours did).

Reply:
Moved to the test interface InspectableDelaying and so no longer available by default.